File: app/src/main/java/knf/kuma/favorite/FavoriteViewModel.java
package knf.kuma.favorite;

import android.content.Context;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import knf.kuma.commons.PrefsUtil;
import knf.kuma.database.CacheDB;
import knf.kuma.pojos.FavoriteObject;

public class FavoriteViewModel extends ViewModel {

    public LiveData<List<FavoriteObject>> getData(Context context) {
        if (PrefsUtil.INSTANCE.showFavSections())
            return FavSectionHelper.init();
        else
            switch (PrefsUtil.INSTANCE.getFavsOrder()) {
                default:
                case 0:
                    return CacheDB.INSTANCE.favsDAO().getAll();
                case 1:
                    return CacheDB.INSTANCE.favsDAO().getAllID();
            }
    }

}

File: app/src/main/java/knf/kuma/favorite/FavoriteViewModel.kt
package knf.kuma.favorite

import androidx.lifecycle.ViewModel
import knf.kuma.commons.PrefsUtil
import knf.kuma.database.CacheDB
import knf.kuma.pojos.av1.FavoriteAV1
import knf.kuma.pojos.av1.FavoriteBase
import knf.kuma.pojos.av1.FavoriteSectionAV1
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.mapLatest

class FavoriteViewModel : ViewModel() {

    val favoriteListFlow = PrefsUtil.favsOrderFlow.flatMapLatest { sortType ->
        CacheDB.INSTANCE.favoriteAV1DAO().byCategoryFlow.mapLatest {
            if (PrefsUtil.showFavSections()) {
                val map = mutableMapOf<String, MutableList<FavoriteAV1>>()
                it.forEach { item ->
                    val list = map[item.category] ?: mutableListOf()
                    list.add(item)
                    map[item.category] =
                        (if (sortType == 1) list.sortedBy { it.aid } else list.sortedBy { it.name }).toMutableList()
                }
                val list = mutableListOf<FavoriteBase>()
                map.forEach {
                    list.add(FavoriteSectionAV1(it.key))
                    list.addAll(it.value)
                }
                list
            } else {
                if (sortType == 1) it.sortedBy { it.aid } else it.sortedBy { it.name }
            }
        }
    }

}
